oeps/oep-0009.rst

```diff
@@ -0,0 +1,315 @@
+                return queryset
+            return queryset.filter(perms[permission](user))
+
+Such a class would be used in a view's `filter_backends`_ attribute or
+could be used by default for all view classes which don't override it.
+
+.. _BasePermission: http://www.django-rest-framework.org/api-guide/permissions/#custom-permissions
+.. _filter_backends: http://www.django-rest-framework.org/api-guide/filtering/#setting-filter-backends
+
+Rationale
+=========
+
+Discussions about authorization in Open edX have made slow progress
+to date because they often got bogged down in the details of which
+particular rules to use for making authorization checks (or at least what
+kinds of rules to use, e.g. role membership) and how to pass the information
+needed to make authorization decisions across service boundaries.  In the
+meantime, working code has needed to make authorization decisions and in the
+absence of concrete guidance has usually resorted to explicit User attribute
+checks, with the corresponding problems outlined in the Motivation section
+above.
+
+Meanwhile, the implementation of `OEP-3`_ has required the ability to apply
+authorization checks in a reusable application which should have no knowledge
+of the exact business logic used to implement them.  To facilitate this and
+break down the overall authorization topic into more manageable chunks, this
+OEP was initiated to handle just best practices of how to be able to perform
+authorization checks in a consistent manner that makes no unreasonable
+assumptions about their implementation.  Here are some of the goals which have
+shaped the recommendations:
+
+* Maintain compatibility with Django admin, Django REST Framework, and other
+  3rd-party Django packages (many of them use Django's authorization API, and
+  a few even use object-level permissions)
+* Reuse existing libraries when feasible
+* Keep a clear separation between the implementation and usage of permissions
+* Don't require database migrations or data loads each time a new permission
+  is added or the implementation of one is changed
+
+.. _OEP-3: https://open-edx-proposals.readthedocs.io/en/latest/oeps/oep-0003.html
+
+
+Backward Compatibility
+=======================
+
+The ``rules`` package can be added to existing packages with minimal impact,
+as it doesn't inherently change the outcome of any authorization checks.  In
+the handful of places where Open edX code actually uses the Django
+authorization API already, the permission implementation can be switched to
+use ``rules`` when convenient (at which point whichever implementation is
+already in use should be deactivated so they don't come to disagree over
+time.)  New code can be written to follow the guidelines in this OEP, and
+existing code can be gradually updated as the need arises.  Effort should
+be made to update all relevant code related to a new explicit permission,
+however, to avoid tempting developers into thinking they've fully updated the
```

Comment:
Nit: This `however` might be better placed at the beginning of the sentence. It took me a few reads to parse the whole sentence due to the `however` in the middle.